Add Cloudwatch metrics integration (#2740)
grafana/agent: docs/sources/configuration/integrations/cloudwatch-exporter-config.md

[`YACE`](https://github.com/nerdswords/yet-another-cloudwatch-exporter/). This allows the collection
of [AWS CloudWatch](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/WhatIsCloudWatch.html) metrics.

This integration allows one to scrape CloudWatch metrics is a set of configurations which we will call `jobs`. There are 
two kind of jobs: [`discovery`](#discovery_job) and [`static`](#static_job).

Configuration reference:

```yaml
  autoscrape:
    # Enables autoscrape of integrations.
      [ enable: <boolean> | default = true ]

      # Specifies the metrics instance name to send metrics to. Instance
      # names are located at metrics.configs[].name from the top-level config.
      # The instance must exist.
      #
      # As it is common to use the name "default" for your primary instance,
      # we assume the same here.
      [ metrics_instance: <string> | default = "default" ]

      # Autoscrape interval and timeout. Defaults are inherited from the global
      # section of the top-level metrics config.
      [ scrape_interval: <duration> | default = <metrics.global.scrape_interval> ]

>       [ scrape_timeout: <duration> | default = <metrics.global.scrape_timeout> ]

Reviewer: I think you pulled the v2 common config reference, I believe the OOTB params are https://github.com/grafana/agent/blob/e2d6f973e57cd5e863520560b1f5699585a444e7/pkg/integrations/config/config.go#L12-L27

I think the ES docs cover it https://github.com/grafana/agent/blob/e2d6f973e57cd5e863520560b1f5699585a444e7/docs/sources/configuration/integrations/elasticsearch-exporter-config.md#L20-L60

Author: indeed. Fixing!

EC2 instances:

```yaml
sts_region: us-east-2
discovery:
  jobs:
    - type: AWS/EC2
      regions:
        - us-east-2
      metrics:
        - name: CPUUtilization
          period: 5m
          statistics:
            - Average
        - name: NetworkPacketsIn
          period: 5m
          statistics:
            - Average
```

Configuration reference:

```yaml
  # List of AWS regions.
  regions: [ <string> ]

  # List of IAM roles to assume. Defaults to the role on the environment configured AWS role.
  roles: [ <aws_role> ]

  # Cloudwatch service alias ("alb", "ec2", etc) or namespace name ("AWS/EC2", "AWS/S3", etc). See section below for all 
  # supported.
  type: <string>

  # List of `Key/Value` pairs to use for tag filtering (all must match). Value can be a regex.
  seach_tags: [ <aws_tag> ]

  # Custom tags to be added as a list of `Key/Value` pairs. When exported to Prometheus format, the label name follows
  # the following format: `custom_tag_{Key}`.
  custom_tags: [ <aws_tag> ]

  # List of metric definitions to scrape.
  metrics: [ <metric> ] 
```

# static_job

> A static job allows one to scrape an individual CloudWatch metric. For that, them metrics needs to be fully qualified. That

Reviewer 2: ```suggestion
A static job allows one to scrape an individual CloudWatch metric. For that, metrics needs to be fully qualified, specifying the following:
```

> means, the following attributes of a CW metric need to be specified in order for the agent to find it:

Reviewer 2: ```suggestion
```
```suggestion

```

require `Resource`, `Service`, `Class`, and `Type` dimensions to be specified. Same applies to CloudWatch custom metrics,
all dimensions attached to a metric when saved in CloudWatch are required.

Configuration reference:

```yaml
  # List of AWS regions.
  regions: [ <string> ]

  # List of IAM roles to assume. Defaults to the role on the environment configured AWS role.
  roles: [ <aws_role> ]

  # Identifier of the static scraping job. When exported to Prometheus format corresponds to the `name` label.
  name: <string>

  # CloudWatch namespace
  namespace: <string>

  # CloudWatch metric dimensions as a list of Name/Value pairs. Must uniquely define a single metric.
  dimensions: [ <aws_dimension> ]

  # Custom tags to be added as a list of Key/Value pairs. When exported, the label name follows the following format:
  # `custom_tag_{Key}`.
  custom_tags: [ <aws_tag> ]

  # List of metric definitions to scrape.
  metrics: [ <metric> ] 
```

# aws_role

Represents an [AWS IAM Role](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles.html). Required when configuring a job. If omitted
the AWS role that the credentials configured in the environment posses will be used. 

> Useful when scraping metrics from different AWS accounts with a single pair of credentials. In this case, a different role

Reviewer 2: ```suggestion
This is useful when scraping metrics from different AWS accounts with a single pair of credentials. In this case, a different role
```

is configured for the agent to assume prior to calling AWS APIs, therefore, the credentials configured in the system need
permission to assume the target role. See [this documentation](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_use_permissions-to-switch.html) on how to configure this.

```yaml
  # AWS IAM Role ARN the exporter should assume to perform AWS API calls.
  role_arn: <string>

  # External ID used when starting an STS session.
  external_id: <string>
```

# aws_dimension

Represents an [AWS CloudWatch Dimension](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/cloudwatch_concepts.html#Dimension).

```yaml
  name: <string>
  value: <string>
```

# aws_tag

Represents an [AWS Tag](https://docs.aws.amazon.com/general/latest/gr/aws_tagging.html).

```yaml
  key: <string>
  value: <string>
```

# metric

```yaml
  # CloudWatch metric name.
  name: <string>
  
  # List of statistic types, e.g. "Minimum", "Maximum", etc.
  statistics: [ <string> ]
  
  # See the `Period` section below.
  period: [ <duration> | default = 5m ]
```

# Period

> Period controls how much back in time CloudWatch metrics are considered, during each agent scrape. We can split how these 

Reviewer 2: ```suggestion
Period controls how far back in time CloudWatch metrics are considered, during each agent scrape. We can split how these 
```

```
"apigateway:GET"
```

The following IAM permissions are required to discover tagged [Database Migration Service](https://aws.amazon.com/dms/) (DMS) replication instances and tasks:
```
"dms:DescribeReplicationInstances",
"dms:DescribeReplicationTasks"
```

For simplicity, one can use the following AWS IAM Policy for all features of this integration to work.

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Sid": "Stmt1674249227793",
      "Action": [
        "tag:GetResources",
        "cloudwatch:GetMetricData",
        "cloudwatch:GetMetricStatistics",
        "cloudwatch:ListMetrics",
        "ec2:DescribeTags",
        "ec2:DescribeInstances",
        "ec2:DescribeRegions",
        "ec2:DescribeTransitGateway*",
        "apigateway:GET",
        "dms:DescribeReplicationInstances",
        "dms:DescribeReplicationTasks"
      ],
      "Effect": "Allow",
      "Resource": "*"
    }
  ]
}
```

# Supported services in discovery jobs

The following is a list of AWS services that are supported in `cloudwatch_exporter` discovery jobs. When configuring a

> discovery job, the `type` field of each `discovery_job` must match either one of the desired job namespace or alias.

Reviewer 2: ```suggestion
discovery job, the `type` field of each `discovery_job` must match either the desired job namespace or alias.
```

---
aliases:

- /docs/agent/latest/configuration/integrations/cloudwatch-exporter/ title: cloudwatch_exporter

---

# cloudwatch_exporter_config

The `cloudwatch_exporter_config` block configures the `cloudwatch_exporter` integration, which is an embedded version of
[`YACE`](https://github.com/nerdswords/yet-another-cloudwatch-exporter/). This allows the collection
of [AWS CloudWatch](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/WhatIsCloudWatch.html) metrics.

> This integration allows one to scrape CloudWatch metrics is a set of configurations which we will call `jobs`. There are 

Reviewer 3: ```suggestion
This integration allows one to scrape CloudWatch metrics in a set of configurations which we will call `jobs`. There are 
```

two kind of jobs: [`discovery`](#discovery_job) and [`static`](#static_job).

Configuration reference:

```yaml
  #
  # Common Integration Settings
  #
    
  # Enables the elasticsearch_exporter integration, allowing the Agent to automatically

>   # collect system metrics from the configured ElasticSearch server address

Reviewer 3: Elasticsearch also appears later in this file.

>   # collect system metrics from the configured ElasticSearch server address

Author: 🤦  fixing

  # Allows for relabeling labels on the target.
  relabel_configs:
    [- <relabel_config> ... ]

  # Relabel metrics coming from the integration, allowing to drop series
  # from the integration that you don't care about.
  metric_relabel_configs:
    [ - <relabel_config> ... ]

  # How frequent to truncate the WAL for this integration.
  [wal_truncate_frequency: <duration> | default = "60m"]

  #
  # Exporter-specific configuration options
  #

  # Required: AWS region to use when calling STS (https://docs.aws.amazon.com/STS/latest/APIReference/welcome.html) for retrieving
  # account information.
  # Ex: us-east-2
  sts_region: <string>

  discovery:

    # Optional: List of tags (value) per service (key) to export in all metrics. For example defining the ["name", "type"] under
    # AWS/EC2 will export the name and type tags and its values as labels in all metrics. Affects all discovery jobs.
    # Ex:
    # exported_tags:
    #   AWS/EC2:
    #     - name
    exported_tags:
      { <string>: [ <string> ] }

    # List of discovery jobs
    jobs: [ <discovery_job> ]

  # List of static jobs
  static: [ <static_job> ]
```

> # discovery_job

Reviewer 4: Are all these sections supposed to be at this heading level (`#`)?

Author: Nice catch! Fixing

  name: <string>
  
  # Required: List of statistic types, e.g. "Minimum", "Maximum", etc.
  statistics: [ <string> ]
  
  # Optional: See the `Period` section below.
  period: [ <duration> | default = 5m ]
```

### Period

Period controls how far back in time CloudWatch metrics are considered, during each agent scrape. We can split how these 
settings affects the produced values in two different scenarios.

If all metrics within a job (discovery or static) have the same `Period` value configured, CloudWatch APIs will be requested
for metrics from the scrape time, to `Periods` seconds in the past. The values of these are exported to Prometheus.

![](../assets/single-period-time-model.png)

On the other hand, if metrics with different `Periods` are configured under an individual job, this works differently.
First, two variables are calculated aggregating all periods: `length`, taking the maximum value of all periods, and 
the new `period` value, taking the minimum of all periods. Then, CloudWatch APIs will be requested for metrics from
`now - length` to `now`, aggregating each in samples for `period` seconds. For each metrics, the most recent sample
is exported to CloudWatch.

> ![](../assets/multiple-period-time-model.png)

Reviewer 5: This image has .png.png as suffix. Should rename image file. 

There's also a new method for storing images on S3 described in the Writers' Toolkit: https://grafana.com/docs/writers-toolkit/writing-guide/image-guidelines/#where-to-store-media-assets.
